File: 2024/day_25.ts
function execute(code: string): number {
  let cnt = 0;

  const runCode = (block: string): void => {
    let i = 0;
    while (i < block.length) {
      const curr = block[i];
      if (curr === '+') {
        cnt++;
      } else if (curr === '-') {
        cnt--;
      } else if (curr === '[') {
        const end = block.indexOf(']', i);
        const newCode = block.slice(i + 1, end);

        while (cnt !== 0) {
          runCode(newCode);
        }

        i = end;
      } else if (curr === '{') {
        if (cnt === 0) {
          i = block.indexOf('}', i);
        }
      }

      i++;
    }
  };

  code = code.replaceAll('>', '');
  runCode(code);

  return cnt;
}

File: 2024/day_25.py
def execute(code: str) -> int:
  cnt = 0

  def run_code(block: str):
    nonlocal cnt

    i = 0
    while i < len(block):
      if block[i] == "+":
        cnt += 1
      elif block[i] == "-":
        cnt -= 1
      elif block[i] == "[":
        end = block.find("]", i)
        new_code = block[i + 1:end]

        while cnt != 0:
          run_code(new_code)

        i = end
      elif block[i] == "{":
        if cnt == 0:
          i = block.find("}", i)

      i += 1

  code = code.replace('>', '')
  run_code(code)

  return cnt
